### PR TITLE
Adds dispenser, take two

### PR DESCRIPTION
**Pull request by SplinterGP**
## About The Pull Request
Adds a Dispenser for the engineer, currently it ~heals~(salves and bandages, direct heals robotic limbs) nearby humans and allows them to carry alot of items in, might make it do other stuff on future, its equipped on backpack slot
image

## Why It's Good For The Game
Lets engineers drag around alot of items with them without being immediately accessible and also gives them a small healing utility(similar to tricordrazine but only burn and brute).

## Changelog
:cl: Coding By Splinter and Sprite by Zaha Sanko
add: Added the engineer dispenser which is equipped on back slot and carries alot of items, can only heal and access items when deployed
imageadd: new title screen announcing the new engineer dispenser
/:cl:

### CODE REVIEW

**Comment by SplinterGP** [comment]
image

**Comment by Luxyboo** [comment]
id say this would be pretty good for when primos are fully in.

**Comment by ilikesaying** [comment]
Engineers start receiving empty welders and yellow pipe cleaners after the new dispensers are out

**Comment by ZahaSanko** [comment]
where credit

**Comment by iLikeMemes** [comment]
If you didnt know there is actually a game called TerraTech, I highly doubt this will cause any kind of copyright shenanigans but just a heads up. (https://store.steampowered.com/app/285920/TerraTech/) besides that, based PR

File: code/__DEFINES/equipment.dm [hunk 1]
```diff
@@ -67,9 +67,10 @@
 #define IS_DEPLOYABLE (1<<13) //Item can be deployed into a machine
 #define DEPLOY_ON_INITIALIZE (1<<14)
 #define IS_DEPLOYED (1<<15) //If this is on an item, said item is currently deployed
-#define DEPLOYED_NO_PICKUP  (1<<16) //Disables deployed item pickup
-#define DEPLOYED_NO_ROTATE  (1<<17) //Disables deployed item rotation abilities to rotate.
-#define DEPLOYED_WRENCH_DISASSEMBLE (1<<18) //If this is on an item, the item can only be disassembled using a wrench once deployed.
+	#define IS_DEPLOYING (1<<16) //If this is on an item, said item is currently deploying
+#define DEPLOYED_NO_PICKUP  (1<<17) //Disables deployed item pickup
```
Comment (TiviPlus) [inline code/__DEFINES/equipment.dm:71]: why indent

File: code/game/objects/items/storage/backpack.dm [hunk 2]
```diff
@@ -777,3 +778,143 @@
...
+/obj/item/storage/backpack/dispenser
+	name = "Dispense-O-Matic 9000 Provisions Dispenser"
+	desc = "The Dispense-O-Matic also known as \"Dispenser\" is a machine capable of holding a big amount of items on it, while also healing nearby humans. Your allies will often ask you to lay down one of those."
+	icon = 'icons/obj/items/storage/storage_48.dmi'
+	icon_state = "dispenser"
+	flags_equip_slot = ITEM_SLOT_BACK
+	max_w_class = 6
+	max_storage_space = 63
+	COOLDOWN_DECLARE(deploy_cooldown)
+	var/list/affecting_list
```
Comment (TiviPlus) [inline code/game/objects/items/storage/backpack.dm:791]: autodoc

File: code/game/objects/items/storage/backpack.dm [hunk 3]
```diff
@@ -777,3 +778,143 @@
...
+	anchored = TRUE
+	dir = REVERSE_DIR(user.dir)
+	icon_state = "dispenser_deployed"
+	balloon_alert_to_viewers("Deploying...")
+	flick("dispenser_deploy", src)
+	ENABLE_BITFIELD(flags_item, IS_DEPLOYING)
+
+	addtimer(CALLBACK(src, .proc/deploy), 4.2 SECONDS)
+
+/obj/item/storage/backpack/dispenser/proc/deploy()
```
Comment (TiviPlus) [inline code/game/objects/items/storage/backpack.dm:840]: ditto, do the rest as well

File: code/game/objects/items/storage/backpack.dm [hunk 4]
```diff
@@ -777,3 +778,143 @@
...
+	if(!CHECK_BITFIELD(flags_item, IS_DEPLOYING))
+		return
+	DISABLE_BITFIELD(flags_item, IS_DEPLOYING)
+	ENABLE_BITFIELD(flags_item, IS_DEPLOYED)
+	affecting_list = list()
+	for(var/mob/living/carbon/human/human in view(3))
+		if(!line_of_sight(src, human, 2))
+			continue
```
Comment (TiviPlus) [inline code/game/objects/items/storage/backpack.dm:848]: >view
>checks view again

File: code/game/objects/items/storage/backpack.dm [hunk 5]
```diff
@@ -777,3 +778,143 @@
...
+		RegisterSignal(human, COMSIG_PARENT_QDELETING, .proc/on_affecting_qdel)
+		affecting_list[human] = beam(human, "blood_light")
+		human.playsound_local(get_turf(src), 'sound/machines/dispenser/dispenser_heal.ogg', 50)
+	for(var/turf/turfs in range(2, src))
+		RegisterSignal(turfs, COMSIG_ATOM_ENTERED, .proc/entered_tiles)
```
Comment (TiviPlus) [inline code/game/objects/items/storage/backpack.dm:853]: range_turf and AS

File: code/game/objects/items/storage/backpack.dm [hunk 6]
```diff
@@ -777,3 +778,143 @@
...
+	START_PROCESSING(SSprocessing, src)
+
+/obj/item/storage/backpack/dispenser/proc/on_affecting_qdel(datum/source)
+	affecting_list -= source
```
Comment (TiviPlus) [inline code/game/objects/items/storage/backpack.dm:857]: Signal handlers are missing

File: code/game/objects/items/storage/backpack.dm [hunk 7]
```diff
@@ -777,3 +778,143 @@
...
+		for(var/turf/turfs in range(2, src))
+			UnregisterSignal(turfs, COMSIG_ATOM_ENTERED)
+		for(var/mob/living/carbon/human/affecting AS in affecting_list)
+			qdel(affecting_list[affecting])
+			UnregisterSignal(affecting, COMSIG_PARENT_QDELETING)
+		affecting_list = null
+		STOP_PROCESSING(SSobj, src)
+		addtimer(CALLBACK(src, .proc/undeploy), 4.2 SECONDS)
+		return
+	return ..()
```
Comment (TiviPlus) [inline code/game/objects/items/storage/backpack.dm:877]: early return

File: code/game/objects/items/storage/backpack.dm [hunk 8]
```diff
@@ -777,3 +778,143 @@
...
+
+/obj/item/storage/backpack/dispenser/proc/undeploy()
+	if(CHECK_BITFIELD(flags_item, IS_DEPLOYED))
+		density = FALSE
```
Comment (TiviPlus) [inline code/game/objects/items/storage/backpack.dm:881]: both early return and autodoc

File: code/game/objects/items/storage/backpack.dm [hunk 9]
```diff
@@ -777,3 +778,143 @@
...
+					break
+				continue
+			if(limb.is_bandaged() && limb.is_salved())
+				continue
+			limb.bandage()
+			limb.salve()
+			break
+
+
+/obj/item/storage/backpack/dispenser/proc/entered_tiles(datum/source, mob/living/carbon/human/entering)
```
Comment (TiviPlus) [inline code/game/objects/items/storage/backpack.dm:910]: bla bla autodoc im not gonna do it for the rest

**Review by TiviPlus — CHANGES_REQUESTED** [review]
Remove the salving, robot heal is ok

Also please remove the following copyrighted/trademarked content
- Valve/TF2 poster
- Terratech references (if these are on sprites/code)

**Comment by Pill-U** [comment]
How do you get one? I'd assume from engi prep / req but I don't see it

**Comment by novaepee** [comment]
If this gets merged, I might just make this a title painting.

**Comment by SplinterGP** [comment]
> If this gets merged, I might just make this a title painting.

already made

File: code/game/objects/items/storage/backpack.dm (hunk 4 above)
Comment (SplinterGP) [inline code/game/objects/items/storage/backpack.dm:848]:
```suggestion
```

File: code/game/objects/items/storage/backpack.dm (hunk 6 above)
Comment (SplinterGP) [inline code/game/objects/items/storage/backpack.dm:857]:
```suggestion
/obj/item/storage/backpack/dispenser/proc/on_affecting_qdel(datum/source)
	SIGNAL_HANDLER
	affecting_list -= source
```

File: code/game/objects/items/storage/backpack.dm (hunk 7 above)
Comment (SplinterGP) [inline code/game/objects/items/storage/backpack.dm:877]:
```suggestion
	if(CHECK_BITFIELD(flags_item, IS_DEPLOYED) && !CHECK_BITFIELD(flags_item, IS_DEPLOYING))
		return ..()
	if(!can_interact(user))
		return
	balloon_alert_to_viewers("Undeploying...")

	icon_state = "dispenser"
	flick("dispenser_undeploy", src)
	ENABLE_BITFIELD(flags_item, IS_DEPLOYING)
	for(var/turf/turfs in range(2, src))
		UnregisterSignal(turfs, COMSIG_ATOM_ENTERED)
	for(var/mob/living/carbon/human/affecting AS in affecting_list)
		qdel(affecting_list[affecting])
		UnregisterSignal(affecting, COMSIG_PARENT_QDELETING)
	affecting_list = null
	STOP_PROCESSING(SSobj, src)
	addtimer(CALLBACK(src, .proc/undeploy), 4.2 SECONDS)
```

File: code/game/objects/items/storage/backpack.dm [hunk 10]
```diff
@@ -777,3 +778,143 @@
...
+		addtimer(CALLBACK(src, .proc/undeploy), 4.2 SECONDS)
+		return
+	return ..()
+
+/obj/item/storage/backpack/dispenser/proc/undeploy()
+	if(CHECK_BITFIELD(flags_item, IS_DEPLOYED))
+		density = FALSE
+		anchored = FALSE
+		DISABLE_BITFIELD(flags_item, IS_DEPLOYED)
+		DISABLE_BITFIELD(flags_item, IS_DEPLOYING)
```
Comment (SplinterGP) [inline code/game/objects/items/storage/backpack.dm:884]:
```suggestion
/obj/item/storage/backpack/dispenser/proc/undeploy()
	if(CHECK_BITFIELD(flags_item, IS_DEPLOYED))
		return
	density = FALSE
	anchored = FALSE
	DISABLE_BITFIELD(flags_item, IS_DEPLOYED)
	DISABLE_BITFIELD(flags_item, IS_DEPLOYING)
```

File: code/__DEFINES/equipment.dm [hunk 11]
```diff
@@ -67,9 +67,10 @@
 #define IS_DEPLOYABLE (1<<13) //Item can be deployed into a machine
 #define DEPLOY_ON_INITIALIZE (1<<14)
 #define IS_DEPLOYED (1<<15) //If this is on an item, said item is currently deployed
-#define DEPLOYED_NO_PICKUP  (1<<16) //Disables deployed item pickup
-#define DEPLOYED_NO_ROTATE  (1<<17) //Disables deployed item rotation abilities to rotate.
-#define DEPLOYED_WRENCH_DISASSEMBLE (1<<18) //If this is on an item, the item can only be disassembled using a wrench once deployed.
+	#define IS_DEPLOYING (1<<16) //If this is on an item, said item is currently deploying
```
Comment (SplinterGP) [inline code/__DEFINES/equipment.dm:70]:
```suggestion
#define IS_DEPLOYING (1<<16) //If this is on an item, said item is currently deploying
```

**Comment by SplinterGP** [comment]
>     * Terratech references (if these are on sprites/code)

Nope, those are just on the comment for funs.

**Comment by SplinterGP** [comment]
will do autodoc soon

**Comment by TiviPlus** [comment]
image

File: code/__DEFINES/equipment.dm [hunk 12]
```diff
@@ -67,9 +67,10 @@
 #define IS_DEPLOYABLE (1<<13) //Item can be deployed into a machine
 #define DEPLOY_ON_INITIALIZE (1<<14)
 #define IS_DEPLOYED (1<<15) //If this is on an item, said item is currently deployed
-#define DEPLOYED_NO_PICKUP  (1<<16) //Disables deployed item pickup
-#define DEPLOYED_NO_ROTATE  (1<<17) //Disables deployed item rotation abilities to rotate.
-#define DEPLOYED_WRENCH_DISASSEMBLE (1<<18) //If this is on an item, the item can only be disassembled using a wrench once deployed.
+#define IS_DEPLOYING (1<<16) //If this is on an item, said item is currently deploying
```
Comment (BraveMole) [inline code/__DEFINES/equipment.dm:70]: should be a trait

File: code/game/objects/items/storage/backpack.dm [hunk 13]
```diff
@@ -14,6 +14,7 @@
 	max_storage_space = 24
 	access_delay = 1.5 SECONDS
 
+
```
Comment (BraveMole) [inline code/game/objects/items/storage/backpack.dm:17]: .

File: code/game/objects/items/storage/backpack.dm [hunk 14]
```diff
@@ -777,3 +778,137 @@
...
+/obj/item/storage/backpack/dispenser
+	name = "TX-9000 Provisions Dispenser"
+	desc = "The TX-9000 also known as \"Dispenser\" is a machine capable of holding a big amount of items on it, while also healing nearby humans. Your allies will often ask you to lay down one of those."
+	icon = 'icons/obj/items/storage/storage_48.dmi'
+	icon_state = "dispenser"
+	flags_equip_slot = ITEM_SLOT_BACK
+	max_w_class = 6
+	max_storage_space = 63
+	///list of stuff we're currently affecting in our area.
```
Comment (BraveMole) [inline code/game/objects/items/storage/backpack.dm:790]: of stuff? Need a cast

File: code/game/objects/items/storage/backpack.dm [hunk 15]
```diff
@@ -777,3 +778,137 @@
...
+		return
+	user.balloon_alert(user, "You start deploying...")
+
+	user.temporarilyRemoveItemFromInventory(src)
+
+	forceMove(deploy_location)
+	density = TRUE
+	anchored = TRUE
+	dir = REVERSE_DIR(user.dir)
+	icon_state = "dispenser_deployed"
```
Comment (BraveMole) [inline code/game/objects/items/storage/backpack.dm:830]: should be done in update_icon_state

File: code/game/objects/items/storage/storage.dm [hunk 16]
```diff
@@ -527,12 +528,12 @@
...
-			open(user)
-	else
-		. = ..()
-		for(var/mob/M in content_watchers)
-			close(M)
+			return
+		else if(open(user))
+			return
+	. = ..()
+	for(var/mob/M in content_watchers)
```
Comment (BraveMole) [inline code/game/objects/items/storage/storage.dm:535]: AS

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/game/objects/items/storage/storage.dm (hunk 16 above)
Comment (SplinterGP) [inline code/game/objects/items/storage/storage.dm:535]:
```suggestion
	for(var/mob/M AS in content_watchers)
```

File: code/game/objects/items/storage/backpack.dm [hunk 17]
```diff
@@ -777,3 +778,137 @@
...
+	name = "TX-9000 Provisions Dispenser"
+	desc = "The TX-9000 also known as \"Dispenser\" is a machine capable of holding a big amount of items on it, while also healing nearby humans. Your allies will often ask you to lay down one of those."
+	icon = 'icons/obj/items/storage/storage_48.dmi'
+	icon_state = "dispenser"
+	flags_equip_slot = ITEM_SLOT_BACK
+	max_w_class = 6
+	max_storage_space = 63
+	///list of stuff we're currently affecting in our area.
+	var/list/affecting_list
+
```
Comment (TiviPlus) [inline code/game/objects/items/storage/backpack.dm:792]: cast this and define "stuff" above

File: code/game/objects/items/storage/backpack.dm [hunk 18]
```diff
@@ -777,3 +778,137 @@
...
+		if(over_object == usr && ishuman(over_object))
+			open(over_object)
+		return
+	return ..()
+
+/obj/item/storage/backpack/dispenser/attack_self(mob/user)
+	if(!ishuman(user) || CHECK_BITFIELD(flags_item, NODROP))
+		return
+	var/deploy_location = get_step(user, user.dir)
+	if(check_blocked_turf(deploy_location))
```
Comment (TiviPlus) [inline code/game/objects/items/storage/backpack.dm:816]: no parentcall?

**Review by TiviPlus — CHANGES_REQUESTED** [review]

**Comment by theta006** [comment]
PR going stale i cry

**Comment by SplinterGP** [comment]
going to unstale it soon, just got busy with some stuff

**Pull request closed without merging**